### PR TITLE
[AST] NFC: Formalize Decl validation tracking via RAII

### PR DESCRIPTION
**Pull request by davezarzycki**
After this change, RAII ensures that the validation state is accurate as possible.

### CODE REVIEW

**Comment by slavapestov** [comment]
I like the idea of using RAII to set some of these new flags, but I’m not a fan of the renaming.

**Comment by davezarzycki** [comment]
What would you propose then for the names? The problem as I see it was twofold:
1) Before this change, `isBeingValidated()` returned false before validation was actually done. This was done to enable name lookup, etc for later validation logic. I tried to clarify what was going on with the new API names. Do you have any better naming ideas for this scenario?
2) `setValidationStarted()` is misleading name that dates back to when there was an actual "validation started" bit which no longer exists. In practice, the API is used to mark decls as valid. Would you prefer `setValidated()`? Something else?

**Comment by slavapestov** [comment]
> Before this change, isBeingValidated() returned false before validation was actually done. This was done to enable name lookup, etc for later validation logic.

I'm not sure that's why. The "is being validated" flag exists to catch re-entrant calls to validateDecl() from taking place. If the declaration is already being validated or has completed validation, we want to return. Otherwise, we can begin validation.

The "validation completed" state is distinct from "has interface type" because type declarations get an interface type before getting a generic signature.

**Comment by davezarzycki** [comment]
The failure looks unrelated.

@swift-ci please clean smoke test

**Comment by davezarzycki** [comment]
Ping.

**Comment by davezarzycki** [comment]
Ping. Are there any changes that are wanted?

**Comment by davezarzycki** [comment]
Ping.

**Comment by davezarzycki** [comment]
Ping. It has been almost a month. Should I abandon this pull request?

**Comment by davezarzycki** [comment]
Hi @DougGregor – If you could be so kind, how should I interpret the lack of responses to "pings"?

@swift-ci please test

**Comment by davezarzycki** [comment]
@slavapestov and @DougGregor  – Ping. Feedback would be appreciated. Thanks!

**Comment by slavapestov** [comment]
@davezarzycki What is the concrete benefit of this change? Are you using the new validation state for anything? In the long term, we would like to switch everything over to use the request evaluator, and validateDecl(), together with these state bits, will go away entirely. They're too imprecise and error prone.

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -4727,36 +4721,37 @@ void TypeChecker::validateDeclForNameLookup(ValueDecl *D) {
...
-          typealias->setInvalid();
-          typealias->getUnderlyingTypeLoc().setInvalidType(Context);
-        }
+          // Note that this doesn't set the generic environment of the alias yet,
+          // because we haven't built one for the protocol.
+          //
+          // See how validateDecl() sets the generic environment on alias members
+          // explicitly.
+          //
+          // FIXME: Hopefully this can all go away with the ITC.
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: The ITC is now gone. Our hopes and dreams are pinned on the request-evaluator ;)

**Comment by davezarzycki** [comment]
I agree that the current validation tracking is imprecise and error prone. That is why I thought that RAII seemed like a natural way of trending in the right direction. If the validation tracking can go away, that's even better. What do you want to do in the short to medium term?

For the record, I'm not using the improved tracking in this pull request for feature work. Just better/easier compiler debugging.

**Comment by DougGregor** [comment]
The use of RAII here is adding more discipline to a messy area. While it's true that we expect this mess of state to go away with the request-evaluator (once we've peeled off enough requests that `validateDecl` can effectively go away), that could take awhile... weeks at least, months probably. I'm fine with merging this PR as an incremental improvement to the status quo.

**Review by DougGregor — APPROVED** [review]

**Pull request merged**